File: python/chapter5/prog.py
from operator import itemgetter
import random


def sort_by_priority(a, p):
    pairs = list(zip(a, p))
    pairs.sort(key=itemgetter(1))
    for i in range(len(a)):
        a[i] = pairs[i][0]


def permute_by_sorting(a):
    n = len(a)
    p = [1] * n
    for i in range(n):
        p[i] = random.randint(1, n ** 3)
    sort_by_priority(a, p)

File: go/chapter05/prog.go
package chapter05

import (
	"math"
	"math/rand"
	"sort"
	"time"
)

type pair struct {
	value    float64
	priority int
}

func sortByPriority(a []float64, p []int) {
	n := len(a)
	pairs := make([]pair, n)
	for i := 0; i < n; i++ {
		pairs[i] = pair{a[i], p[i]}
	}
	sort.SliceStable(pairs, func(i, j int) bool {
		return pairs[i].priority < pairs[j].priority
	})
	for i := 0; i < n; i++ {
		a[i] = pairs[i].value
	}
}

func permuteBySorting(a []float64) {
	n := len(a)
	p := make([]int, n)
	rand.Seed(time.Now().UnixNano())
	for i := 0; i < n; i++ {
		p[i] = rand.Intn(int(math.Pow(float64(n), 3)))
	}
	sortByPriority(a, p)
}
